implementation of discontinue_on for products + variants

jasonfb opened the pull request:
@JDutil / community -- this should be complete now. If there aren't any blockers to this getting in, please merge for 3.1. If you see any other reasons this should not move forward, please let me know. 

Abbreviated description (full notes below) has bene added to the release notes. 

---

fixes to migration
fixes specs in spree-core
adding discontinue_on to permitted attributes
adds discontinued to admin interface
rake task to fix orphan line items
changing naming of field to 'discontinue_on' for parity with existing field 'available_on'
adding not_discontinued to search scopes for product
changing not_deleted in active scope to not_discontinued
fixes migration problem
adds view & search of discontinued products to Admin -> Products screen
adding to 3.1 release ntoes
displays error message in admin when deleting a product not allowed for deletion

---

Spree's current implementation uses act_as_paranoid to implement a ActiveRecord-wide "soft-delete" of Products and Variants. Unfortunately, soft deleting means that the records are left in the database but behave as if they are really deleted.

Because associations from other objects (like line itmes to variant) won't normally see the deleted, core code is forced (unnaturally) to use scopes .with_deleted 

https://github.com/spree/spree/blob/master/core/app/models/spree/product.rb#L224

Not only does this abstract magic get in the way of acts_as_paranoid's true intention, but  in a worse example, it becomes necessary to overload an AR belongs_to with a magic method.

https://github.com/spree/spree/blob/master/core/app/models/spree/line_item.rb#L106

This hack above solves the problem that when looking at a past order with line items attached to deleted variants you can't load the variants because the objects appear to be missing in Active Record.

The design flaw here is that these objects really should not be considered "deleted." The word is wrong, the concept is wrong, and the implementation bends over backwards to accomodate the flawed design. 

The approach proposed solves the underlying flaw _and all the related bugs cuased by this flaw_ in the following ways:
- [x] Adds a field discontinue_on to Products and Variants. 
- [x] Migrate the timestamps deleted_at to discontinue_on (when possible), and un-delete the deleted variants (when there is not matching SKU) and products (when there is no matching slug)
- [x] Redefine the available scope on the Products object as so. Notice that the not_discontinued scope is added to the end of whatever scope chain is here. This is necessary because we no longer have the default_scope weeding out discontinued products:
  
  ```
    add_search_scope :not_discontinued do
      where("#{Product.quoted_table_name}.discontinue_on IS NULL or #{Product.quoted_table_name}.discontinue_on >= ?", Time.zone.now)
    end
    search_scopes << :not_discontinued
  
    # Can't use add_search_scope for this as it needs a default argument
    def self.available(available_on = nil, currency = nil)
      not_discontinued.joins(:master => :prices).where("#{Product.quoted_table_name}.available_on <= ?", available_on || Time.now)
    end
    search_scopes << :available
  
    def self.active(currency = nil)
      not_discontinued.available(nil, currency)
    end
  ```
- [x] Add a `not_discontinued` scope on the Variant object:
  
  ```
    scope :not_discontinued, -> { where("#{Variant.quoted_table_name}.discontinue_on IS NULL OR #{Variant.quoted_table_name}.discontinue_on <= ?",  Time.now) }
  ```

(the active class method has been changed to include this scope)
- [x] Removes all references to unscope association chains from other objects to the Product & Variant objects in places where unscope is used explicitly to work-around the default scope problem. (This is a big win because it makes the associations cleaner and easier to work with.)
- [x] Finally, although it is slightly counter-intuitive, I have actually left the deleted_at fields in place (although their data will be moved to discontinued_at field and their values will be reset to NULL in the db migration). You can (and should!) use deleting to remove human-error mistakes (real mistakes) before the items get sold in your store, or in the case when you have duplicate slugs (Products) or SKUs (Variants) in your database. In those special cases only, you should continue to use delete. In all other cases, use the new discontinue_on feature. 
- [x] You can only delete a Product or Variant if no orders have been placed for that product/variant. Once the variant is associated with a Line item, it can never be "deleted," and instead you must use the new discontinue_on feature. Model-level checks (`before_destroy`) enforce this.
- [x] Note: The DB migration should fix your database correctly unless you have created new Products & Variants with matching slugs/SKUs of deleted records. In this case, you must use the included `rake db:fix_orphan_line_items` task to clean up your records. Both the schema migration and the script are very pro-active in helping you fix your own database. 
- [x] Not only do we fix _all_ product & variant `undefined method XYZ for nil` bugs (and all the bugs yet-to-be-created), but we also get a new feature: You can now schedule a product or variant to be discontinued in the future. Neato!
- [x] The Admin interface now shows a distinction between discontinued & deleted, as well as options for searching for discontinued products. 
- [x] update Spree's release notes https://github.com/spree/spree/blob/master/guides/content/release_notes/3_1_0.md
- [x] When I delete a Product or Variant that has associated line items, fail gracefully in admin

Commit:
basic implementation of discontinued_at for products + variants
fixes to migration
fixes specs in spree-core
adding discontinue_on to permitted attributes
adds discontinued to admin interface
rake task to fix orphan line items
changing naming of field to 'discontinue_on' for parity with existing field 'available_on'
adding not_discontinued to search scopes for product
changing not_deleted in active scope to not_discontinued
fixes migration problem
adds view & search of discontinued products to Admin -> Products screen
adding to 3.1 release ntoes
displays error message in admin when deleting a product not allowed for deletion

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://jasonfb-add-discontinued-at-to-products-and-variants2.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://jasonfb-add-discontinued-at-to-products-and-variants2.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

JDutil commented on backend/app/controllers/spree/admin/products_controller.rb:
```diff
@@ -44,9 +44,14 @@ def update
 
-        flash[:success] = Spree.t('notice_messages.product_deleted')
+        begin
+          # TODO: why is @product.destroy raising ActiveRecord::RecordNotDestroyed instead of failing with false result
```
weird

JDutil replied:
Is this a acts_as_paranoid bug?

JDutil commented on backend/app/views/spree/admin/shared/_destroy.js.erb:
```diff
@@ -3,4 +3,9 @@ if success %>
 <% end %>
 
+<% error = flash.discard(:error)
+if error %>
```
Think this could be `if error = flash.discard(:error)`

JDutil commented on backend/spec/spec_helper.rb:
```diff
@@ -47,7 +47,7 @@
 
 # Set timeout to something high enough to allow CI to pass
-Capybara.default_wait_time = 10
+Capybara.default_wait_time = 35
```
I don't think this should be so high.  Maybe 15 or 30 tops most servers like unicorn will timeout at 30 seconds, and once a request times out once there tends to be several so this will take much longer time to get a final build result.

jasonfb replied:
Sorry I swear my MacBook Air can't finish the suite without raising these. I had meant to *not* check these in, I will revert

JDutil replied:
I'm okay with making it 30 that's the timeout of most web servers.

dhonig replied:
you running build.sh ?  my air has no issues with it that way

On Mon, Sep 14, 2015 at 6:04 PM, Jason Fleetwood-Boldt <
notifications@github.com> wrote:

> In backend/spec/spec_helper.rb
> <https://github.com/spree/spree/pull/6731#discussion_r39453394>:
>
> > @@ -47,7 +47,7 @@
> >  Capybara.javascript_driver = :poltergeist
> >
> >  # Set timeout to something high enough to allow CI to pass
> > -Capybara.default_wait_time = 10
> > +Capybara.default_wait_time = 35
>
> Sorry I swear my MacBook Air can't finish the suite without raising these.
> I had meant to *not* check these in, I will revert
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree/spree/pull/6731/files#r39453394>.
>

JDutil commented on core/db/migrate/20150819154308_add_discontinued_to_products_and_variants.rb:
```diff
@@ -0,0 +1,66 @@
+class AddDiscontinuedToProductsAndVariants < ActiveRecord::Migration
+  def up
+    add_column :spree_products, :discontinue_on, :datetime, after: :available_on
```
these columns should have indexes

JDutil commented on guides/content/release_notes/3_1_0.md:
```diff
@@ -29,6 +29,27 @@ For more information, see the [taxation guide](https://guides.spreecommerce.com/
+
+We are fixing this by adding new feilds 'Discontinue On' to products & variants (discontinue_on)
+                                                        
+This fixes a design flaw in that in mos stores these objects really should not be considered "deleted."  The approach proposed solves the underlying flaw and all the related bugs cuased by this flaw in the following ways:
```
in most stores*

JDutil commented:
@jasonfb left some minor feedback, and would be good to cleanup hound warnings as well. Other than that it looks good to me.

Commit:
syntax fix

Commit:
reverting Capybara.default_wait_time change

Commit:
reverting timeout change

Commit:
adding indexes to discontinue_on fields

Commit:
fixing hound warnigns & spelling

Commit:
hound warnings

Commit:
hound warnings

Commit:
hound warnings

Commit:
Fixing hound warnings

jasonfb commented:
@JDutil -- I addressed the feedback (Except the Hound line-too-long since all of those were on text that I can't add linebreaks to)

TODO: need to squash & rebase with master. If you could please re-review to see if there are another changes you'd like, I'd appreciate it.

JDutil commented:
After rebasing I think it'll be okay

JDutil commented on guides/content/release_notes/3_1_0.md:
```diff
@@ -29,6 +29,27 @@ For more information, see the [taxation guide](https://guides.spreecommerce.com/
 Spree Alert model has been removed, and will no longer ping alerts.spreecommerce.com
 to check for notifications.
 
+### Future Discontinue of Products & Vairiants
```
Variants*

jasonfb commented:
replaced with https://github.com/spree/spree/pull/6794
